Auto migrate the version table

goetas opened the pull request:
This is a proposal on how to migrate the v2 versions table to v3 versions.

The thing that worries me here is that we do not ask the user if we should alter the table or not.

Does it make sense to have a separate command for it?

goetas committed:
use multiple migrations directories and namespaces

goetas committed:
update more commands

goetas committed:
increase version

goetas committed:
allow custom sorting

goetas committed:
executor

goetas committed:
remove start migration

goetas committed:
value object comparison for versions

goetas committed:
adapted execute command

goetas committed:
do not load from sub namespaces

goetas committed:
format migration state

goetas committed:
start adapting tests

goetas committed:
stricter execution result

goetas committed:
create MigratorConfigurationFactory

goetas committed:
extract factory interface

goetas committed:
fix event dispatcher

goetas committed:
use configuration loaders, drop inheritance

goetas committed:
add config loader tests

goetas committed:
config tests

goetas committed:
executor tests

goetas committed:
command tests

goetas committed:
up to date command

goetas committed:
migrations repo tests

goetas committed:
metadata tests

goetas committed:
alias resolver tests

goetas committed:
factory test

goetas committed:
cs and tests

goetas committed:
interface

goetas committed:
autocommit listener

goetas committed:
migration plan calculator tests

goetas committed:
simplify code for migration plan calculator

goetas committed:
more tests

goetas committed:
rollup test

goetas committed:
restore file builder

goetas committed:
test schema diff

goetas committed:
stub schema provider

goetas committed:
dump schema test

goetas committed:
generate test

goetas committed:
diff command

goetas committed:
dix deps

goetas committed:
execute command

goetas committed:
migrate command

goetas committed:
remove outdated tests

goetas committed:
simplify class

goetas committed:
Update documentation

goetas committed:
more changes

goetas committed:
add event arg test

goetas committed:
static analysis

goetas committed:
moved new-migrations and unavailable migrations to the plan calculator

goetas committed:
green tests

goetas committed:
fix most of the phpstan stuff

goetas committed:
update xsd validator for 3.0

goetas committed:
phpstan

goetas committed:
add constructor provided migrations

goetas committed:
add missing range feature

goetas committed:
fix some failing tests

goetas committed:
missing factory method

goetas committed:
add config loader

goetas committed:
first/last tests

goetas committed:
di test

goetas committed:
move migrator option to command

goetas committed:
migrator tests

goetas committed:
phpstan and tests

goetas committed:
restore down migration behaviour

goetas committed:
stof code review suggestions

goetas committed:
use datetime immutable

goetas committed:
mark value objects as final

goetas committed:
single queries are run in debug mode

goetas committed:
ensure proper typecasting

goetas committed:
skip migration version table in the schema dumps

goetas committed:
allow to filter tables in a dump

goetas committed:
prepare upgrading document

goetas commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -137,4 +137,67 @@ public function complete(ExecutionResult $result) : void
+
+        $comparator   = new Comparator();
+        $currentTable = $this->schemaManager->listTableDetails($this->configuration->getTableName());
+        $diff         = $comparator->diffTable($currentTable, $expectedSchemaChangelog);
```
This part relies on doctrine schema diff to update the table.

goetas commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -137,4 +137,67 @@ public function complete(ExecutionResult $result) : void
+        return $this->schemaManager->tablesExist([$this->configuration->getTableName()]);
+    }
+
+    private function updateMigratedVersions() : void
```
This function performs the update of migrated versions from the `YYYYMMDDHHIISS` format to FQCN

goetas committed:
version it is a string

goetas committed:
fix too large column size

goetas committed:
change output levels

goetas commented:
ping @alcaeus

alcaeus commented:
> The thing that worries me here is that we do not ask the user if we should alter the table or not.

I'm not sure if a separate command on its own would be the solution.

For most commands, we can ask an interactive question if we see that there's a diff to the migrations table. If the users approves, we update their table and then continue the migration normally. If they don't want to upgrade, we abort the migration.

For commands that are run non-interactively, e.g. as part of a deployment, we obviously can't get any consent, so I believe the best course of action would be to tell people to run a separate command to update manually, then restart the migration process.

goetas committed:
fixed a lot of stuff from code review

goetas committed:
move generateVersionNumber out of config class

goetas committed:
rename MigratorConfigurationFactory

goetas committed:
rename Factory

goetas committed:
cs

goetas committed:
add version equals method

goetas committed:
rename interfaces

goetas committed:
more readable executed at code

goetas committed:
ensure lowercase column names

goetas committed:
document migrator interface

goetas committed:
code reviews by @greg0ire

goetas committed:
alphabetical sorting of loaders

goetas committed:
change argument order of ConnectionHelperLoader

goetas committed:
rename PhpFileLoader

goetas committed:
remove param default

goetas committed:
mark getDirectoryRelativeToFile as final

goetas committed:
rename UnableToLoadResource

goetas committed:
mark ConfigurationLoader as internal

goetas committed:
split table storage documentation

goetas committed:
MigrationNotAvailable::forVersion

goetas committed:
update composer lock

goetas committed:
handle possible regex errors

goetas committed:
diff command tests fixed after rebase

goetas committed:
upgrade

goetas committed:
run cs on php 7.2

goetas committed:
windows tests

goetas committed:
filesystem sorting is not relevant in the finder

goetas committed:
migration plan execution is immutable

goetas committed:
warn to not use registerMigrationInstance

goetas committed:
regex finder accepts any class naming

goetas committed:
do not allow invalid config keys

goetas committed:
array loader is a hard dependency for the other loaders

goetas committed:
no invalid keys are allowed

goetas committed:
refer to https://github.com/Roave/BackwardCompatibilityCheck instead of having a long list

OskarStark commented:
> Does it make sense to have a separate command for it?

I would say yes 👍

goetas commented:
There are the following options: 

1) Do silently the conversion on the migrations_versions table

- pro: simple and convenient, no extra actions are needed from the user
- con: implicitly writing to the database (a write command will be issued on a simple `migration:status`  command as example

2) Have a separate migration command to update the migrations_versions, if another command is issued and the table is not up to date, return an error

- pro: simple to implement
- con: now we introduce a new step to be performed. Probably will annoy who uses automated processes in CI/CD

3) Add a `--allow-metadata-update` parameter to all the commands. Apply option N.1 if the param is not specified. Return an error if the parameter is not specified and the table is not up to date. Still offer a command for manually migrating the table (meas offer N2 too)

- pro: no unexpected behaviors that the user is not aware of (everything is explicit)
- con: a bit more stuff to implement
- mixed pro/cons:  new `--allow-metadata-update` to be added on CI/CD pipelines 


I'm personally for option n.3. Any other idea?

alcaeus commented:
I would agree that method 3 is the safest choice, as it will require people to acknowledge the change and decide whether they want to individually approve every version table migration or allow it to be updated automatically

goetas committed:
classgenerator test

goetas committed:
add commands to migrate table structure

goetas committed:
test executed migrations update

goetas commented:
- added checks in "read" commands that detect if the storage is not up to date
- added command to update metadata storage
- auto-migrating metadata storage when running migrations
- added tests for related updates of schema and data

goetas commented:
Moving to https://github.com/doctrine/migrations/pull/878
